crates/legacy/src/testing/finalization_test.rs

```diff
@@ -0,0 +1,550 @@
+use std::{num::NonZeroUsize, sync::Arc, time::Duration};
+
+use crate::{
+    builder_state::{DaProposalMessage, QuorumProposalMessage},
+    service::{GlobalState, ProxyGlobalState, ReceivedTransaction},
+    BuilderStateId, ParentBlockReferences,
+};
+use async_broadcast::{broadcast, Sender};
+use async_lock::RwLock;
+use committable::Commitment;
+use hotshot::{
+    traits::BlockPayload,
+    types::{BLSPubKey, SignatureKey},
+};
+use hotshot_builder_api::{
+    v0_2::{block_info::AvailableBlockInfo, data_source::BuilderDataSource},
+    v0_3::{builder::BuildError, data_source::AcceptsTxnSubmits},
+};
+use hotshot_example_types::{
+    block_types::{TestBlockHeader, TestBlockPayload, TestMetadata, TestTransaction},
+    node_types::{TestTypes, TestVersions},
+    state_types::{TestInstanceState, TestValidatedState},
+};
+use hotshot_types::{
+    data::{DaProposal, QuorumProposal, ViewNumber},
+    message::Proposal,
+    simple_certificate::QuorumCertificate,
+    traits::{
+        block_contents::{vid_commitment, BlockHeader},
+        node_implementation::ConsensusTime,
+    },
+    utils::BuilderCommitment,
+};
+use sha2::{Digest, Sha256};
+
+use super::basic_test::{BuilderState, MessageType};
+
+type TestSetup = (
+    ProxyGlobalState<TestTypes>,
+    async_broadcast::Sender<MessageType<TestTypes>>,
+    async_broadcast::Sender<MessageType<TestTypes>>,
+    async_broadcast::Sender<MessageType<TestTypes>>,
+    async_broadcast::Sender<Arc<ReceivedTransaction<TestTypes>>>,
+);
+
+const TEST_NUM_NODES_IN_VID_COMPUTATION: usize = 4;
+const TEST_NUM_CONSENSUS_RETRIES: usize = 4;
+
+/// [setup_builder_for_test] sets up a test environment for the builder state.
+/// It returns a tuple containing the proxy global state, the sender for decide
+/// messages, the sender for data availability proposals,
+fn setup_builder_for_test(allow_empty_block_period: Option<u64>) -> TestSetup {
+    let (req_sender, req_receiver) = broadcast(32);
+    let (tx_sender, tx_receiver) = broadcast(32);
```

Comment:
Any reason that the number of messages are set to 32? Would it be useful if we define it as a const?

Reply:
It's entirely arbitrary in this case.  But this is a good point.

Addressed in 11cfcf20619527dd4e46555dce64316d4516c8dd.